lib/metrics/queue.js

```diff
@@ -0,0 +1,85 @@
+'use strict'
+class QueueMetricsCollector {
+  constructor () {
+    this.stats = {
+      'queue.latency.min.ms': 0,
+      'queue.latency.max.ms': 0,
+      'queue.latency.avg.ms': 0
+    }
+
+    this.total = 0
+    this.count = 0
+    this.min = 0
+    this.max = 0
+  }
+
+  // Updates the data used to generate our stats
+  //
+  // Unlike the Stats and RuntimeCollector, this function
+  // also returns the instantiated collector.
+  //
+  // @param {number} time A javascript ms timestamp
+  updateStats (time) {
+    if (this.min === 0 || this.min > time) {
+      this.min = time
+    }
+    if (this.max < time) {
+      this.max = time
+    }
+
+    this.count++
+    this.total += time
+  }
+
+  // standard `collect` method for stats collectors
+  //
+  // called by the MetricReporter object prior to its sending data
+  //
+  // @param {function} cb callback function
+  collect (cb) {
+    // update average based on count and total
```

Comment:
Given https://github.com/elastic/apm-agent-nodejs/issues/1977 (and its root https://github.com/elastic/kibana/pull/90955) should we be:

1. Avoiding reporting metrics for this period if count==0? Currently, once a queue metric is reported for a particular queue, it will be reported every interval, even if there is no further activity on that queue.
2. Consider changing `const queueMetics = new Map()` in sqs.js to be a data structure that evicts entries if they haven't been used in a long time. Currently if a long running app cycles through lots of queue names over time, then we'll add (small) mem usage for those over time.  This data structure could be accomplished via https://github.com/isaacs/node-lru-cache (using max=Infinity, and maxAge=...some timeout value...).

The former (1) is a bigger issue.

For example, I setup my app to use queue "trentm-play-sqs3" for a while, then stop using and use queue "trentm-play-sqs2". The reported metrics every interval show empty metrics for the now unused "trentm-play-sqs3":

```
    {"metricset":{"samples":{"queue.latency.min.ms":{"value":825300},"queue.latency.max.ms":{"value":3382475},"queue.latency.avg.ms":{"value":1163652.6875}},"timestamp":1617651611908000,"tags":{"queue_name":"trentm-play-sqs2","hostname":"pink.local","env":"development"}}}
    {"metricset":{"samples":{"queue.latency.min.ms":{"value":0},"queue.latency.max.ms":{"value":0},"queue.latency.avg.ms":{"value":0}},"timestamp":1617651611908000,"tags":{"queue_name":"trentm-play-sqs3","hostname":"pink.local","env":"development"}}}
```

Reply:
Good questions -- I see three.  Working backwards: 

Re: `node-lru-cache` or something similar -- that feels like a premature optimization.  A user would need to have a significant number of queues running and instrumented for this to be noticeable.  If node had a built-in data structure that expired the least recently used items or if the agent already has a system for doing this I'd say we go for it now. I don't think (at this time) it's worth the cost of a new third party dependency, or building out or own "expire the old stuff" memory cache. If we see evidence SQS instrumentations create high memory problems we can/should revisit.

Reply:
Re: skipping a metric if the count is zero -- that's another good question without a clear win tradeoff wise.  To take the other side of the argument a measurement of 0 is still a measurement that might provide insight into how a system is running, and the absence of a metric isn't necessarily a value of 0.  If we _are_ going to change the behavior of reporting 0-value metrics, it seems like the right place to do that is when metrics [are serialized and reported](https://github.com/elastic/apm-agent-nodejs/blob/a2a6152c8009887efec4b10e29f4ac47b7d7dab2/lib/metrics/reporter.js#L18), and I'd prefer to do it separate from this PR.

Reply:
Re: the high cardinality problem, we're less susceptible that that here.  The problem Kibana ran into is they had [a slightly pathological case](https://elastic.slack.com/archives/C5M02D1GE/p1612974533191700?thread_ts=1612965746.189300&cid=C5M02D1GE) 

> e.g. there’s an endpoint that’s basically `/:buildHash/bundles/:plugin/:filename`, which is called for every asset loaded from the browser, and per file/build hash/plugin it generates a unique transaction name

_Every_ frontend asset served (in dev mode)

    ex. https://kibana.example.com:9243/39309/bundles/plugin/mapsLegacy/mapsLegacy.chunk.1.js
    
was creating a unique transaction name when the file was first served.  I presume that's because, in dev mode, the static files are actually served/bundled by node.js on the fly as requests for them come in.  On the home page this can be _literally_ hundreds of different transaction names.  

This bad problem was made worse by [the transaction breakdown metrics](https://github.com/elastic/apm-agent-nodejs/blob/a2a6152c8009887efec4b10e29f4ac47b7d7dab2/lib/instrumentation/transaction.js#L276).  These metrics include transaction name as a metric dimension (tag/label).  So hundreds, possibly thousands, of unique metric sets were created and then stuck around.

The only extra metric dimension/label/tag we use here is queue name.  It's important to capture this metric as it allows filtering.  A single application/service would need to be a very large number of queues for this to become a perf. issue.  So, again, I think the right trade-off here is to launch for the 80% case, and if the 20% cases start rolling in we can address it then.  Regardless of when we address it, it feels like a separate PR that targets metric reporting generally.